### PR TITLE
Let M68kMCCodeEmitter set Scratch size.

### PR DESCRIPTION
**Pull request by erikfjon**
The Scratch buffer passed to getBinaryCodeForInst needs to be able to
hold any value returned by getMachineOpValue or other custom encoders.
It's better to let the caller of getBinaryCodeForInst set the size of
Scratch as it's impossible for VarLenCodeEmitterGen to know what the
smallest needed size is.

VarLenCodeEmitterGen now calculates its smallest needed Scratch bit
width based on the slice operations and zero extends Scratch if it's too
small. This only guarantees that Scratch has enough bits for the
generated code not for getMachineOpValue or custom encoders.

The smallest internal APInt representation uses one uint64_t word so
there is no point in using a smaller size.

### CODE REVIEW

**Comment by erikfjon** [comment]
I have locally verified that `check-llvm-mc-m68k` and `check-llvm-codegen-m68k` passes.

**Comment by erikfjon** [comment]
Updated commit message with missing tags `[Tablegen]` and `[M68k]`.

File: llvm/utils/TableGen/VarLenCodeEmitterGen.cpp [hunk 1]
```diff
@@ -505,7 +506,16 @@ std::string VarLenCodeEmitterGen::getInstructionCaseForEncoding(
   if (!PostEmitter.empty())
     SS.indent(I) << "Inst = " << PostEmitter << "(MI, Inst, STI);\n";
 
-  return Case;
+  // Resize the scratch buffer if it's to small.
+  std::string ScratchResizeStr = "";
+  if (VLI.size() && !VLI.isFixedValueOnly()) {
+    raw_string_ostream RS(ScratchResizeStr);
+    std::string High = utostr(HighScratchAccess);
```
Comment (mshockwave) [inline llvm/utils/TableGen/VarLenCodeEmitterGen.cpp:513]: I think `raw_string_ostream` can convert unsigned values to strings. So maybe
```
<< "if (Scratch.getBitWidth() < " << High << ") "
```
would be better than calling `utostr` here.

File: llvm/utils/TableGen/VarLenCodeEmitterGen.cpp [hunk 2]
```diff
@@ -505,7 +506,16 @@ std::string VarLenCodeEmitterGen::getInstructionCaseForEncoding(
   if (!PostEmitter.empty())
     SS.indent(I) << "Inst = " << PostEmitter << "(MI, Inst, STI);\n";
 
-  return Case;
+  // Resize the scratch buffer if it's to small.
+  std::string ScratchResizeStr = "";
```
Comment (mshockwave) [inline llvm/utils/TableGen/VarLenCodeEmitterGen.cpp:510]: nit: the default ctor of `std::string` already gives you an empty string.

**Review by mshockwave — COMMENTED** [review]
This is a pretty neat optimization as the original version might allocated more bits than we needed for `Scratch`. Thank you!
I only have a few comments.

**Review by mshockwave — APPROVED** [review]
LGTM

**Pull request merged**